### DISCUSSION TITLE
[Feautre request] Allow arbitrary values for defineProperties function in sprinkle package

### DISCUSSION THREAD

**Discussion by joseDaKing** [vanilla-extract-css/vanilla-extract#550] [Ideas]
Allow arbitrary values for defineProperties function in sprinkle package by using functions. This would allow for more flexibility. These functions should also be able to take variables from createVar function.

```ts
import { defineProperties } from "@vanilla-extract/sprinkles";

const tokenVars = {
    space: {
        xs: "128px",
        md: "256px"
    }
};

const mapToCSSProp = <T extends keyof CSSProperties>(cssProp: T) => (value: CSSProperties[T]) => value;

const mapToThemeProp = <T extends keyof CSSProperties, K>(cssProp: T, themeValues: K) => (value: CSSProperties[T] | keyof K) => {
    return (themeValues as any)[value] ?? value;
};

const space = defineProperties({
    conditions,
    defaultCondition: "_",
    properties: {
        justifyContent: mapToCSSProp("justifyContent"),
        margin: mapToThemeProp("margin", tokenVars.space)
    }
});

const spaceSprinkle = createSprinkles(space);

spaceSprinkle({
  // Only one class should be generated for each arbitrary value
  margin: "32px"
});

spaceSprinkle({
  margin: "xs" // Would be 128px in css
});
```

**Comment by graup** [comment]
Not sure how this would work with arbitrary props (like your `mapToCSSProp`). If the values are not known at compile time, how can they be converted to classes or variables?

The variable example (`mapToThemeProp`) seems useful. I actually had a similar use case where I wanted to map variants to variables. This is a button recipe where a variant uses the `color` in different ways:

```ts
import { createVar } from '@vanilla-extract/css';
import { recipe } from '@vanilla-extract/recipes';

export const currentColor = createVar();

export const buttonRecipe = recipe({
  base: {
    vars: {
      [currentColor]: 'grey',
    },
  },

  variants: {
    variant: {
      stroke: { boxShadow: `0 0 1px 1px ${currentColor}` },
      'strong-fill': { backgroundColor: currentColor, color: '#fff' },
    },
  },

  defaultVariants: {
    variant: 'stroke',
  },
});
```

For now, the only way to do this seems to be using `assignInlineVars` to translate a `color` prop to the variable. Would be nice if I could automatically translate a prop to a variant based on a variable without extra code.

```ts
function getProps({ color, variant, ...props }: Props) {
  const className = buttonRecipe({ variant });
  const style = assignInlineVars({[currentColor]: vars.color[var]});
  return mergeProps(props, {className, style});
}
export function Button(props: Props) {
  return <button {...getProps(props)} />;
}

```

I'd love to do something like this in the recipe:
```ts
  variants: {
    color: mapToVar(currentColor, vars.color),
    variant: {
      stroke: { boxShadow: `0 0 1px 1px ${currentColor}` },
      'strong-fill': { backgroundColor: currentColor, color: '#fff' },
    },
  },
```

**Reply by joseDaKing** [reply]
Actually the more I think about this, the arbitrary function should not take variables from createVar then everything could be evaluated at compile-time, it would work more like tailwindcss.

**Reply by didimedina** [reply]
If this were possible this would be a game-changer for me. Ultimately this would allow me to set up a project, create my system based on assumptions of what I'll need, and over time evolve it based on newly discovered patterns that are developing in my codebase. That, and also being able to use one-off styles which is very common on landing pages ie unique CSS grid settings etc.
